### ISSUE TITLE
selftest: ASSERT_STREQ: test/account.c:98: expected = 'sip:[fe80::b62e:99ff:feee:268f]', actual = 'sip:[fe80:b62e:99ff:feee:268f]@proxy.com'

### ISSUE THREAD

**Issue by robert-scheck** [baresip/baresip#1771]
Running `make check` fails randomly on Fedora 34, 35, 36 and 37 (Rawhide) only on the ppc64le hardware architecture like this:

```
+ ./selftest
running baresip selftest version 2.0.0 with 46 tests
[ RUN      ] test_account
[       OK ]
[ RUN      ] test_account_uri_complete
selftest: ASSERT_STREQ: test/account.c:98: expected = 'sip:[fe80::b62e:99ff:feee:268f]', actual = 'sip:[fe80:b62e:99ff:feee:268f]@proxy.com'
[;mTEST_ERR: test/account.c:178: [fe80::b62e:99ff:feee:268f] (Bad message)
[;mtest_account_uri_complete: test failed (Bad message)
[;mtest failed (Bad message)
[;mre main loop:
  maxfds:  1024
  nfds:    6
  method:  3 (epoll)
```

Is it possible that some variable inside `selftest` is not properly initialized and thus contains garbage?

 - Fedora 37 ppc64le failure: https://koji.fedoraproject.org/koji/taskinfo?taskID=84100044
 - Fedora 37 ppc64le failure: https://koji.fedoraproject.org/koji/taskinfo?taskID=84100530
 - Fedora 36 ppc64le failure: https://koji.fedoraproject.org/koji/taskinfo?taskID=84100098
 - Fedora 35 ppc64le failure: https://koji.fedoraproject.org/koji/taskinfo?taskID=84100763
 - Fedora 34 ppc64le failure: https://koji.fedoraproject.org/koji/taskinfo?taskID=84100351
 - Fedora 34 ppc64le failure: https://koji.fedoraproject.org/koji/taskinfo?taskID=84099289

However, simply re-running a failed build may lead to this:

- Fedora 37 ppc64le success: https://koji.fedoraproject.org/koji/buildinfo?buildID=1933074
- Fedora 36 ppc64le success: https://koji.fedoraproject.org/koji/buildinfo?buildID=1933077
- Fedora 35 ppc64le success: https://koji.fedoraproject.org/koji/buildinfo?buildID=1933081
- Fedora 34 ppc64le success: https://koji.fedoraproject.org/koji/buildinfo?buildID=1933086

Given the builds take place in a clean chroot environment each time it's unlikely that the build environment is causing this (except for hardware defects, but given the amount of different builders…unlikely).

**Comment by alfredh** [comment]
hi, I have a couple of questions:

- when the test fails, is it the same uri string all the time, or different ones each time ?

- is it possible for you to run valgrind+selftest on that CPU ?

we are already running selftest with valgrind, but this is mainly on the Intel CPU (amd64).
PPC and Intel may have slightly different behaviour.

**Comment by robert-scheck** [comment]
> * when the test fails, is it the same uri string all the time, or different ones each time ?

All build logs are public (linked from the build task linked above), thus:

- https://kojipkgs.fedoraproject.org//work/tasks/44/84100044/build.log: `expected = 'sip:[fe80::b62e:99ff:feee:268f]', actual = 'sip:[fe80:b62e:99ff:feee:268f]@proxy.com'`
- https://kojipkgs.fedoraproject.org//work/tasks/530/84100530/build.log: `expected = 'sip:[fe80::b62e:99ff:feee:268f]', actual = 'sip:[fe80:b62e:99ff:feee:268f]@proxy.com'`
- https://kojipkgs.fedoraproject.org//work/tasks/98/84100098/build.log: `expected = 'sip:[fe80::b62e:99ff:feee:268f]', actual = 'sip:[fe80:b62e:99ff:feee:268f]@proxy.com'`
- https://kojipkgs.fedoraproject.org//work/tasks/763/84100763/build.log: `expected = 'sip:[fe80::b62e:99ff:feee:268f]', actual = 'sip:[fe80:b62e:99ff:feee:268f]@proxy.com'`
- https://kojipkgs.fedoraproject.org//work/tasks/351/84100351/build.log: `expected = 'sip:[fe80::b62e:99ff:feee:268f]', actual = 'sip:[fe80:b62e:99ff:feee:268f]@proxy.com'`

> * is it possible for you to run valgrind+selftest on that CPU ?

Interactively or scripted? The Fedora build system is sealed, thus I only could run it scripted, and the build chroot also has no Internet access (thus valgrind and it's dependencies need to be part of Fedora, which seems to be the case). Do you have an example? I'm not used to valgrind at all…

**Comment by alfredh** [comment]
here is an example:

```sh
$ valgrind --leak-check=yes ./selftest -v test_account_uri_complete
```

it should be enough to run it once on the ppc64le hardware (interactive)...

if the code is fine, it will not print any warnings.
if there are problems with uninited variables etc, it might print some warnings ...

**Comment by alfredh** [comment]
can you also test this patch:

https://github.com/baresip/baresip/pull/1773

**Comment by robert-scheck** [comment]
> ```shell
> valgrind --leak-check=yes ./selftest -v test_account_uri_complete
> ```

```
$ valgrind --leak-check=yes ./selftest -v test_account_uri_complete
==973835== Memcheck, a memory error detector
==973835== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==973835== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==973835== Command: ./selftest -v test_account_uri_complete
==973835== 
running baresip selftest version 2.0.0 with 1 tests
Local network addresses:
       eth0:  10.3.171.53
     virbr0:  192.168.122.1
[ RUN      ] test_account_uri_complete
selftest: ASSERT_STREQ: test/account.c:98: expected = 'sip:[fe80::b62e:99ff:feee:268f]', actual = 'sip:[fe80:b62e:99ff:feee:268f]@proxy.com'
[;mTEST_ERR: test/account.c:178: [fe80::b62e:99ff:feee:268f] (Bad message)
[;mtest_account_uri_complete: test failed (Bad message)
[;mtest failed (Bad message)
[;mre main loop:
  maxfds:  1024
  nfds:    6
  method:  3 (epoll)
ua: stop all (forced=1)
==973835== 
==973835== HEAP SUMMARY:
==973835==     in use at exit: 0 bytes in 0 blocks
==973835==   total heap usage: 4,843 allocs, 4,843 frees, 265,136 bytes allocated
==973835== 
==973835== All heap blocks were freed -- no leaks are possible
==973835== 
==973835== For lists of detected and suppressed errors, rerun with: -s
==973835== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
$ 
```

**Comment by alfredh** [comment]
could you please try again with git HEAD (commit 314b0b4ce606f177e55532e02875e8ac5a806a80) ?

if that also fails, please try to comment out the line in the array with the mentioned uri.

**Comment by robert-scheck** [comment]
> could you please try again with git HEAD (commit [314b0b4](https://github.com/baresip/baresip/commit/314b0b4ce606f177e55532e02875e8ac5a806a80)) ?

```
$ valgrind --leak-check=yes ./selftest -v test_account_uri_complete
==1512336== Memcheck, a memory error detector
==1512336== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==1512336== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==1512336== Command: ./selftest -v test_account_uri_complete
==1512336== 
expected string: (31 bytes)
"sip:[fe80::b62e:99ff:feee:268f]"
actual string: (41 bytes)
"sip:[fe80::b62e:99ff:feee:268f]@proxy.com"
running baresip selftest version 2.0.0 with 1 tests
Local network addresses:
       eth0:  10.3.171.63
     virbr0:  192.168.122.1
[ RUN      ] test_account_uri_complete
TEST_STRCMP: test/account.c:143: failed
[;mtest_account_uri_complete: test failed (Invalid argument)
[;mtest failed (Invalid argument)
[;mre main loop:
  maxfds:  1024
  nfds:    6
  method:  3 (epoll)
ua: stop all (forced=1)
==1512336== 
==1512336== HEAP SUMMARY:
==1512336==     in use at exit: 0 bytes in 0 blocks
==1512336==   total heap usage: 4,817 allocs, 4,817 frees, 258,794 bytes allocated
==1512336== 
==1512336== All heap blocks were freed -- no leaks are possible
==1512336== 
==1512336== For lists of detected and suppressed errors, rerun with: -s
==1512336== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
$ 
```

> if that also fails, please try to comment out the line in the array with the mentioned uri.

```
$ valgrind --leak-check=yes ./selftest -v test_account_uri_complete
==3239803== Memcheck, a memory error detector
==3239803== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==3239803== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==3239803== Command: ./selftest -v test_account_uri_complete
==3239803== 
running baresip selftest version 2.0.0 with 1 tests
Local network addresses:
       eth0:  10.3.171.76
     virbr0:  192.168.122.1
[ RUN      ] test_account_uri_complete
[       OK ]
ua: stop all (forced=1)
OK. 1 tests passed successfully[;m
ua: stop all (forced=1)
==3239803== 
==3239803== HEAP SUMMARY:
==3239803==     in use at exit: 0 bytes in 0 blocks
==3239803==   total heap usage: 4,818 allocs, 4,818 frees, 257,396 bytes allocated
==3239803== 
==3239803== All heap blocks were freed -- no leaks are possible
==3239803== 
==3239803== For lists of detected and suppressed errors, rerun with: -s
==3239803== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
$ 
```

**Comment by alfredh** [comment]
thanks for testing

I think we can ignore the valgrind track now.

I have a feeling that the bug is related to IPv6, if you look closely the machine does not have
an IPv6 address when running the test. It has two IPv4 addresses (eth0, virbr0).

the IPv6 code should work even if the box does not have an IPv6 address,
but perhaps it is not.

if you removed the IPv6 link local address, and the test started working perhaps it is related
to IPv6 link local.

@cspiel1 do you have any ideas ?

**Comment by alfredh** [comment]
I tried to disable ipv6 on my linux box, and the test also failed:

```
$ sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1
```

```
$ ./selftest test_account_uri_complete
running baresip selftest version 2.0.0 with 1 tests
[ RUN      ] test_account_uri_complete
.... input: '192.168.1.2'
.... uri_is_ip: 1
.... input: '192.168.1.2:5677'
.... uri_is_ip: 1
.... input: 'user'
.... uri_is_ip: 0
.... append domain
.... input: 'user@domain.com'
.... uri_is_ip: 0
.... input: 'user@domain.com:5677'
.... uri_is_ip: 0
.... input: 'sip:user'
.... uri_is_ip: 0
.... append domain
.... input: 'sip:user@domain.com'
.... uri_is_ip: 0
.... input: '[2113:1470:1f1b:24b::2]'
.... uri_is_ip: 1
.... input: '[fe80::b62e:99ff:feee:268f]'
.... uri_is_ip: 0
.... append domain

TEST_STRCMP: test/account.c:143: failed
expected string: (31 bytes)
"sip:[fe80::b62e:99ff:feee:268f]"
actual string: (41 bytes)
"sip:[fe80::b62e:99ff:feee:268f]@proxy.com"
test_account_uri_complete: test failed (Invalid argument)
test failed (Invalid argument)
re main loop:
  maxfds:  1024
  nfds:    6
  method:  3 (epoll)

```

the problem is that `sa_set_str` returns an error for IPv6 link local.


There are 2 solutions to ths problem:

1. Disable IPv6 test if the host has no IPv6.
2. Enable IPv6 on the host

**Comment by robert-scheck** [comment]
Commit 0f980ba1d8a12d041c31b9bb4dadd861a37c9c3f doesn't seem to target `test_account_uri_complete`, is that correct?

**Comment by cspiel1** [comment]
- I would prefer to disable the IPv6 test if the host has no IPv6.
- We should be sure that the github action build server has IPv6 support.

**Comment by alfredh** [comment]
@cspiel1 retest also fails if the box has no IPv6 address:

```
test 80 -- test_rtcp_packetloss
test 81 -- test_rtcp_twcc
test 82 -- test_sa_class
test: test_sa_class: test failed (Cannot assign requested address)

retest: TEST_ERR: src/main.c:235: (Cannot assign requested address)
```

could you please have a look ?

**Comment by cspiel1** [comment]
- HAVE_INET6 is set by default, also in a none IPv6 linux box.
- The test fails because it passes a IPv6ll address to `sa_set_str()`.
- disabling HAVE_INET6= was set to "deprecated" by @sreimers (5a8c2c4e3e107bfc1138a842cb652f27241ea19a)
- There was already a discussion to get rid of all the `#ifdef HAVE_INET6`. This would simplify lots of code sections.

Thus IPv6 has to be enabled on the host.

**Comment by alfredh** [comment]
from a user perspective, the program should work if:

1. he has an IPv4 address

2.he has an IPv6 address
3 he has both IPv4 and IPv6 address


Right now the program is failing because he does not have IPv6.
This is not very userfriendly.

The problem is that uri_complete has different behaviour depending on if the host
has IPv6 or not. The culprit is sa_set_str when the input is fe80::xxx Link local address.

I would suggest fixing sa_set_str so that the behaviour is identical, if the host has
IPV4, IPv6 or both.

**Comment by cspiel1** [comment]
Will have a look on `sa_set_str()`.

**Comment by cspiel1** [comment]
Can we be sure that `struct sockaddr_in6` is defined?
If yes, we could remove the `HAVE_INET6` switch in
```c
int sa_set_str(struct sa *sa, const char *addr, uint16_t port)
{
        ...

#ifdef HAVE_INET6
	case AF_INET6:
		sa->u.in6.sin6_port = htons(port);
		sa->len = sizeof(struct sockaddr_in6);
		break;
#endif
        ...
}

```

**Comment by sreimers** [comment]
> Can we be sure that `struct sockaddr_in6` is defined?
I think so, all supported operating systems should be supported since HAVE_INET6 is already the default for many years.
Since we released the deprecation, its fine to drop HAVE_INET6 support (if we move to runtime detection)

**Comment by alfredh** [comment]
here is my proposal for a fix:

https://github.com/baresip/re/pull/297


the problem is twofold:

1. the function account_uri_complete needs to check if the input is a hostname or IP-address.
  for this it is using sa_set_str/sa_pton. The problem can also be fixed by writing a new function
  to determine if a string is hostname/IP-addr.

2. sa_pton is using getaddrinfo for IPv6 linklocal. If the host has no IPv6 address, this will fail.
  inet_pton can convert to/from IPv4/v6 even if the host does not support it.

**Comment by robert-scheck** [comment]
Is https://github.com/baresip/re/pull/297 enough for backporting to Fedora? Or are further commits needed. Or will there be e.g. a re v2.1.2 soon?

**Comment by sreimers** [comment]
> Or will there be e.g. a re v2.1.2 soon?

Just released - re v2.2.0

**Comment by robert-scheck** [comment]
Thank you very much! Builds of re v2.2.0 and baresip v2.0.1 are now succeeding on all active Fedora and EPEL versions and architectures.
